keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,865 @@
+matchConditions:
+  # Exclude node requests from the webhook
+  - expression: '!("system:nodes" in request.userInfo.groups)'
+```
+
+#### Scope an NFS access management webhook to Pods mounting NFS volumes
+
+> I want to narrowly scope my webhook to only the relevant requests, in order to reduce load on the
+> webhook and reduce latency in irrelevant requests.
+
+Concrete example:
+
+> A NFS deployment uses an third-party access management system. I have an admission webhook that
+> performs an access check for against the external system for pods that mount NFS volumes. Only
+> pods with NFS volumes need to be checked.
+
+Currently, there is no way to achieve this. Many webhook implementations today start by checking
+that the request is within scope, and return early if it's not. This adds latency and an additional
+failure point to irrelevant requests. This example requires an external integration, and thus is not
+a candidate for migration to CEL `ValidatingAdmissionPolicy`.
+
+With match conditions, the expressions can check whether the request object is in-scope for the
+webhook:
+
+```yaml
+rules:
+  - operations: ['CREATE']
+    apiGroups: '' # core
+    apiVersions: '*'
+    resources: 'pods'
+matchConditions:
+  # Only include pods with an NFS volume.
+  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
+```
+
+### Goals
+
+1. Provide a filtering mechanism for excluding requests from an admission webhook
+2. Maintain consistency with `ValidatingAdmissionPolicy`
+
+### Non-Goals
+
+* Provide a mechanism to exclude requests from all webhooks.
+
+## Proposal
+
+### API
+
+Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
+with a new `MatchConditions` field:
+
+```go
+
+type ValidatingWebhook struct {
+  // ...
+
+  // MatchConditions is a list of conditions on the AdmissionRequest ('request') that must be met for a
+  // request to be sent to this webhook.
+  // +optional
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+type MutatingWebhook struct {
+  // ...
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+// MatchCondition represents a condition which must by fulfilled for a request to be sent to a webhook.
+type MatchCondition struct {
+	// Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the AdmissionRequest, organized into CEL variables:
+	//
+	//'object' - The object from the incoming request. The value is null for DELETE requests.
+	//'oldObject' - The existing object. The value is null for CREATE requests.
+	//'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// request.object. No other metadata properties are accessible.
```

Comment:
why is access to other metadata fields like labels, annotations, finalizers, ownerReferences restricted?

Reply:
I don't think it should be. 

There's a back story here: This documentation chunk originated in CRD Validation Rules.  For CRD Validation Rules, we did not allow validation of metadata (except for name) to adhere to [rule #4 structural schemas](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema): "if metadata is specified, then only restrictions on metadata.name and metadata.generateName are allowed.".  This rule pre-dates CEL (it is also true for OpenAPIv3 validations) and ensures that metadata is uniform across resources (names being the exception which has it's own back story).

The rationale for CRD Validation Rules does not directly apply to admission rules which historically have been allowed to validate all metadata fields.

Reply:
We need to fix this documentation on ValidatingAdmissionPolicy as well (the implementation does allow access to all metadata fields). I've made a note on https://github.com/kubernetes/enhancements/pull/3717/files#r1085742827 to get this fixed.

Reply:
Ack. I removed most of this comment block, and just put a reference to https://github.com/kubernetes/website/issues/39089 instead.